TiDB panic when run with `tidb_enable_collect_execution_info=0` in TiKV env

Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Run integration test in TiKV env many times, it will panic forever in some rounds. And we will get log like 

```
[2023/11/02 02:13:07.738 +08:00] [ERROR] [coprocessor.go:1117] ["copIteratorWork meet panic"] [r="runtime error: invalid memory address or nil pointer dereference"] ["stack trace"="github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTask.func1
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:1119
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:914
runtime.panicmem
	/usr/local/go/src/runtime/panic.go:261
runtime.sigpanic
	/usr/local/go/src/runtime/signal_unix.go:861
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleCopCache
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:1688
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleCopResponse
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:1437
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTaskOnce
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:1278
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).handleTask
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:1130
github.com/pingcap/tidb/pkg/store/copr.(*copIteratorWorker).run
	/home/jenkins/agent/workspace/pingcap/tidb/periodics_integration_test/tidb/pkg/store/copr/coprocessor.go:817"]
```

### 2. What did you expect to see? (Required)

No panic

### 3. What did you see instead (Required)

panic forever.

### 4. What is your TiDB version? (Required)

master


### Related log link

https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fperiodics_integration_test/detail/periodics_integration_test/171/artifacts/

Log file
[integration-test.txt.zip](https://github.com/pingcap/tidb/files/13237183/integration-test.txt.zip)

--- a/pkg/executor/distsql_test.go
+++ b/pkg/executor/distsql_test.go
@@ -569,3 +569,25 @@ func TestCoprocessorBatchByStore(t *testing.T) {
 		}
 	}
 }
+
+func TestCoprCacheWithoutExecutionInfo(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk1 := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(id int)")
+	tk.MustExec("insert into t values(1), (2), (3)")
+
+	require.NoError(t, failpoint.Enable("github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler/mockCopCacheInUnistore", `return(123)`))
+	defer func() {
+		require.NoError(t, failpoint.Disable("github.com/pingcap/tidb/pkg/store/mockstore/unistore/cophandler/mockCopCacheInUnistore"))
+	}()
+
+	defer tk.MustExec("set @@tidb_enable_collect_execution_info=1")
+	ctx := context.WithValue(context.Background(), "CheckSelectRequestHook", func(_ *kv.Request) {
+		tk1.MustExec("set @@tidb_enable_collect_execution_info=0")
+	})
+	tk.MustQuery("select * from t").Check(testkit.Rows("1", "2", "3"))
+	tk.MustQueryWithContext(ctx, "select * from t").Check(testkit.Rows("1", "2", "3"))
+}

--- a/pkg/store/copr/coprocessor.go
+++ b/pkg/store/copr/coprocessor.go
@@ -1685,7 +1685,13 @@ func (worker *copIteratorWorker) handleCopCache(task *copTask, resp *copResponse
 				resp.pbResp.Range = nil
 			}
 		}
-		resp.detail.CoprCacheHit = true
+		// If the coprocessor cache key of the task is same whether `worker.enableCollectExecutionInfo` is true or false,
+		// the cache may be hit when `worker.enableCollectExecutionInfo` is false, but the `detail` is nil.
+		// Check `resp.detail` to avoid panic.
+		// Details: https://github.com/pingcap/tidb/issues/48212
+		if resp.detail != nil {
+			resp.detail.CoprCacheHit = true
+		}
 		return nil
 	}
 	copr_metrics.CoprCacheCounterMiss.Add(1)

--- a/pkg/store/copr/coprocessor.go
+++ b/pkg/store/copr/coprocessor.go
@@ -1685,8 +1685,9 @@ func (worker *copIteratorWorker) handleCopCache(task *copTask, resp *copResponse
 				resp.pbResp.Range = nil
 			}
 		}
-		// If the coprocessor cache key of the task is same whether `worker.enableCollectExecutionInfo` is true or false,
-		// the cache may be hit when `worker.enableCollectExecutionInfo` is false, but the `detail` is nil.
+		// `worker.enableCollectExecutionInfo` is loaded from the instance's config. Because it's not related to the request,
+		// the cache key can be same when `worker.enableCollectExecutionInfo` is true or false.
+		// When `worker.enableCollectExecutionInfo` is false, the `resp.detail` is nil, and hit cache is still possible.
 		// Check `resp.detail` to avoid panic.
 		// Details: https://github.com/pingcap/tidb/issues/48212
 		if resp.detail != nil {

